tasmota/xdrv_65_ds3502.ino

```diff
@@ -0,0 +1,153 @@
+ * DS3502 - digital potentiometer (https://datasheets.maximintegrated.com/en/ds/DS3502.pdf)
+ *
+ * I2C Address: 0x28 .. 0x2B
+\*********************************************************************************************/
+
+#define XDRV_65                     65
+#define XI2C_65                     65  // See I2CDEVICES.md
+
+ #define DS3502_REG_MODE        0x02
+ #define DS3502_REG_WIPER       0x00
+// #define DS3502_REG_PRE_SCALE       0xFE
+
+#ifndef USE_DS3502_ADDR
+  #define USE_DS3502_ADDR          0x28
+#endif
+
+bool DS3502_detected = false;
+uint8_t DS3502_potiValue;
+
+void DS3502_Detect(void)
+{
+  if (!I2cSetDevice(USE_DS3502_ADDR)) { return; }
+
+  uint8_t buffer;
+  // TODO: init routine
+//   if (I2cValidRead8(&buffer, USE_DS3502_ADDR, DS3502_REG_MODE1)) {
+     I2cWrite8(USE_DS3502_ADDR, DS3502_REG_MODE, 0x80);
+//     if (I2cValidRead8(&buffer, USE_DS3502_ADDR, DS3502_REG_MODE1)) {
+//       if (0x20 == buffer) {
+         DS3502_detected = true;
+         I2cSetActiveFound(USE_DS3502_ADDR, "DS3502");
+         DS3502_Reset(); // Reset the controller
+//       }
+//     }
+//   }
+}
+
+void DS3502_Reset(void)
+{
+  //I2cWrite8(USE_DS3502_ADDR, DS3502_REG_MODE1, 0x80);
+//  DS3502_SetPotifreq(USE_DS3502_FREQ);
+  DS3502_SetPoti(0);
+  DS3502_potiValue = 0;
+  Response_P(PSTR("{\"DS3502\":{\"RESET\":\"OK\"}}"));
+}
+
+void DS3502_SetPoti_Reg(uint8_t val) {
+   I2cWrite8(USE_DS3502_ADDR, DS3502_REG_WIPER, val);
+}
+
+void DS3502_SetPoti(uint8_t poti) {
+  if (127 == poti) {
+    DS3502_SetPoti_Reg(127); // Special use additional bit causes channel to turn on completely without PWM
+  } else {
+    DS3502_SetPoti_Reg(poti);
+  }
+  DS3502_potiValue = poti;
+}
+
+bool DS3502_Command(void)
+{
+  bool serviced = true;
+  bool validpin = false;
+  uint8_t paramcount = 0;
+  if (XdrvMailbox.data_len > 0) {
+    paramcount=1;
+  } else {
+    serviced = false;
+    return serviced;
+  }
+  char argument[XdrvMailbox.data_len];
+  for (uint32_t ca=0;ca<XdrvMailbox.data_len;ca++) {
+    if ((' ' == XdrvMailbox.data[ca]) || ('=' == XdrvMailbox.data[ca])) { XdrvMailbox.data[ca] = ','; }
+    if (',' == XdrvMailbox.data[ca]) { paramcount++; }
+  }
+  UpperCase(XdrvMailbox.data,XdrvMailbox.data);
+
+  if (!strcmp(ArgV(argument, 1),"RESET"))  {  DS3502_Reset(); return serviced; }
+
+  if (!strcmp(ArgV(argument, 1),"STATUS"))  { DS3502_OutputTelemetry(false); return serviced; }
+
+  if (!strcmp(ArgV(argument, 1),"POTI")) {
+    if (paramcount > 1) {
+        uint16_t val = atoi(ArgV(argument, 2));
+        if ((val >= 0 && val <= 127)) {
+          DS3502_SetPoti(val);
+          Response_P(PSTR("{\"DS3502\":{\"POTI\":%i}}"),val);
+          serviced = true;
+          return serviced;
+        }
+    }
+  }
+  return serviced;
+}
+
+void DS3502_OutputTelemetry(bool telemetry)
+{
+  ResponseTime_P(PSTR(",\"DS3502\":{\"POTI\":%i}}"),DS3502_potiValue);
+  if (telemetry) {
+    MqttPublishTeleSensor();
+  }
```

Comment:
This shouldn't be called from within a driver
If you want to consider your poti as a sensor (would be strange) you should implement as a xsns and handle the FUNC_JSON_APPEND function

However I don't see why a poti (which is an output device) should provide telemetry

Reply:
Good point, telemetry is not required. Current state can be checked using status command.